// @shopify/hydrogen/dist/esnext/foundation/Router/BrowserRouter.client.js
import { createBrowserHistory } from 'history';
import React, { createContext, useContext, useMemo, useState, useEffect, } from 'react';
import { META_ENV_SSR } from '../ssr-interop';
import { useServerState } from '../useServerState';
export const RouterContext = createContext({});
let currentPath = '';
let isFirstLoad = true;
export const BrowserRouter = ({ history: pHistory, children, }) => {
    if (META_ENV_SSR)
        return React.createElement(React.Fragment, null, children);
    const history = useMemo(() => pHistory || createBrowserHistory(), [pHistory]);
    const [location, setLocation] = useState(history.location);
    const { pending, serverState, setServerState } = useServerState();
    useEffect(() => {
        // The app has just loaded
        if (isFirstLoad)
            isFirstLoad = false;
        // A navigation event has just happened
        else if (!pending && currentPath !== serverState.pathname) {
            window.scrollTo(0, 0);
        }
        currentPath = serverState.pathname;
    }, [pending]);
    useEffect(() => {
        const unlisten = history.listen(({ location: newLocation }) => {
            setServerState({
                pathname: newLocation.pathname,
                search: location.search || undefined,
            });
            setLocation(newLocation);
        });
        return () => unlisten();
    }, [history]);
    return (React.createElement(RouterContext.Provider, { value: {
            history,
            location,
        } }, children));
};
export function useRouter() {
    const router = useContext(RouterContext);
    if (!router && META_ENV_SSR) {
        throw new Error('useRouter must be used within a <Router> component');
    }
    return router;
}
export function useLocation() {
    return useRouter().location;
}
