// @shopify/hydrogen/dist/node/framework/Hydration/ServerComponentRequest.server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerComponentRequest = void 0;
const timing_1 = require("../../utilities/timing");
const cache_1 = require("../cache");
const react_helmet_async_1 = require("react-helmet-async");
const constants_1 = require("../../constants");
let reqCounter = 0; // For debugging
const generateId = typeof crypto !== 'undefined' &&
    // @ts-ignore
    !!crypto.randomUUID
    ? // @ts-ignore
        () => crypto.randomUUID()
    : () => `req${++reqCounter}`;
// Stores queries by url or '*'
const preloadCache = new Map();
const PRELOAD_ALL = '*';
/**
 * This augments the `Request` object from the Fetch API:
 * @see https://developer.mozilla.org/en-US/docs/Web/API/Request
 *
 * - Adds a `cookies` map for easy access
 * - Adds a static constructor to convert a Node.js `IncomingMessage` to a Request.
 */
class ServerComponentRequest extends Request {
    constructor(input, init) {
        if (input instanceof Request) {
            super(input, init);
        }
        else {
            super(getUrlFromNodeRequest(input), {
                headers: new Headers(input.headers),
                method: input.method,
                body: input.method !== 'GET' && input.method !== 'HEAD'
                    ? input.body
                    : undefined,
            });
        }
        this.time = (0, timing_1.getTime)();
        this.id = generateId();
        this.ctx = {
            cache: new Map(),
            head: new react_helmet_async_1.HelmetData({}),
            router: {
                routeRendered: false,
                serverProps: {},
                routeParams: {},
            },
            queryCacheControl: [],
            queryTimings: [],
            preloadQueries: new Map(),
        };
        this.cookies = this.parseCookies();
        const referer = this.headers.get('referer');
        this.preloadURL =
            this.isRscRequest() && referer && referer !== '' ? referer : this.url;
    }
    parseCookies() {
        const cookieString = this.headers.get('cookie') || '';
        return new Map(cookieString
            .split(';')
            .map((chunk) => chunk.trim().split(/=(.+)/)));
    }
    isRscRequest() {
        const url = new URL(this.url);
        return url.pathname === constants_1.RSC_PATHNAME;
    }
    savePreloadQuery(query) {
        if (typeof query.preload === 'string' && query.preload === PRELOAD_ALL) {
            saveToPreloadAllPreload(query);
        }
        else if (query.preload) {
            this.ctx.preloadQueries.set((0, cache_1.hashKey)(query.key), query);
        }
    }
    getPreloadQueries() {
        if (preloadCache.has(this.preloadURL)) {
            const combinedPreloadQueries = new Map();
            const urlPreloadCache = preloadCache.get(this.preloadURL);
            mergeMapEntries(combinedPreloadQueries, urlPreloadCache);
            mergeMapEntries(combinedPreloadQueries, preloadCache.get(PRELOAD_ALL));
            return combinedPreloadQueries;
        }
        else if (preloadCache.has(PRELOAD_ALL)) {
            return preloadCache.get(PRELOAD_ALL);
        }
    }
    savePreloadQueries() {
        preloadCache.set(this.preloadURL, this.ctx.preloadQueries);
    }
}
exports.ServerComponentRequest = ServerComponentRequest;
function mergeMapEntries(map1, map2) {
    map2 && map2.forEach((v, k) => map1.set(k, v));
}
function saveToPreloadAllPreload(query) {
    let setCache = preloadCache.get(PRELOAD_ALL);
    if (!setCache) {
        setCache = new Map();
    }
    setCache === null || setCache === void 0 ? void 0 : setCache.set((0, cache_1.hashKey)(query.key), query);
    preloadCache.set(PRELOAD_ALL, setCache);
}
/**
 * @see https://github.com/frandiox/vitedge/blob/17f3cd943e86d7c0c71a862985ddd6caa2899425/src/node/utils.js#L19-L24
 *
 * Note: Request can sometimes be an instance of Express request, where `originalUrl` is the true source of what the
 * URL pathname is. We want to use that if it's present, so we union type this to `any`.
 */
function getUrlFromNodeRequest(request) {
    var _a;
    const url = (_a = request.originalUrl) !== null && _a !== void 0 ? _a : request.url;
    if (url && !url.startsWith('/'))
        return url;
    // TODO: Find out how to determine https from `request` object without forwarded proto
    const secure = request.headers['x-forwarded-proto'] === 'https';
    return new URL(`${secure ? 'https' : 'http'}://${request.headers.host + url}`).toString();
}
